graylog2-server/src/main/java/org/graylog2/plugin/Message.java

```diff
@@ -953,4 +962,59 @@ public String apply(final Message input) {
             return input.getId();
         }
     }
+
+    /**
+     * Store the specified metadata value within the message's internal metadata map.
+     *
+     * @param key   A globally unique string key identifier for the metadata value.
+     * @param value The metadata object value.
+     */
+    public void setMetadata(String key, Object value) {
+        Preconditions.checkArgument(StringUtils.isNotEmpty(key), "A non-empty key is required.");
+        Preconditions.checkNotNull(value);
+        if (metadata == null) {
+            metadata = new HashMap<>();
+        }
+        metadata.put(key, value);
+    }
+
+    /**
+     * Get the metadata value for the specified key.
+     *
+     * @param key The string key for the metadata entry.
+     */
+    @Nullable
+    public Object getMetadataValue(String key) {
+        if (metadata == null) {
+            return null;
+        }
+        return metadata.get(key);
+    }
+
+    /**
+     * Get the metadata value for the specified key. If not present, then return the default value.
+     *
+     * @param key The string key for the metadata entry.
+     */
+    @Nullable
+    public Object getMetadataValue(String key, Object defaultValue) {
+        if (metadata == null) {
+            return null;
+        }
```

Comment:
So if no metadata has been set, then an attempt to get-with-default returns `null` instead of the default?  Is there a reason we wouldn't want to return the provided default value?

Reply:
Thanks @waab76 Great point and find. This is a bug. I'll fix it and add test coverage.

Reply:
Thanks for the feedback. Fixed.